### PR TITLE
oneDNN: Enable aarch64 builds

### PR DESCRIPTION
**Pull request by hmartinez82**
```
Hernan@ARM64-VM CLANGARM64 /c/Dev/Github/MINGW-packages/mingw-w64-onednn/src/build-CLANGARM64
$ ninja test
[0/1] Running tests...Test project C:/Dev/Github/MINGW-packages/mingw-w64-onednn/src/build-CLANGARM64
        Start   1: cpu-bnorm-u8-via-binary-postops-cpp
  1/102 Test   #1: cpu-bnorm-u8-via-binary-postops-cpp ..............   Passed    0.06 sec
        Start   2: cpu-cnn-inference-f32-c
  2/102 Test   #2: cpu-cnn-inference-f32-c ..........................   Passed    0.09 sec
        Start   3: cpu-cnn-inference-f32-cpp
  3/102 Test   #3: cpu-cnn-inference-f32-cpp ........................   Passed    5.21 sec
        Start   4: cpu-cnn-inference-int8-cpp
  4/102 Test   #4: cpu-cnn-inference-int8-cpp .......................   Passed    0.16 sec
        Start   5: cpu-cnn-training-bf16-cpp
  5/102 Test   #5: cpu-cnn-training-bf16-cpp ........................   Passed    0.09 sec
        Start   6: cpu-cnn-training-f32-cpp
  6/102 Test   #6: cpu-cnn-training-f32-cpp .........................   Passed    0.54 sec
        Start   7: cpu-cnn-training-f32-c
  7/102 Test   #7: cpu-cnn-training-f32-c ...........................   Passed    1.07 sec
        Start   8: cpu-rnn-inference-f32-cpp
  8/102 Test   #8: cpu-rnn-inference-f32-cpp ........................   Passed    0.12 sec
        Start   9: cpu-rnn-inference-int8-cpp
  9/102 Test   #9: cpu-rnn-inference-int8-cpp .......................   Passed    0.03 sec
        Start  10: cpu-getting-started-cpp
 10/102 Test  #10: cpu-getting-started-cpp ..........................   Passed    0.03 sec
        Start  11: cpu-graph-getting-started-cpp
 11/102 Test  #11: cpu-graph-getting-started-cpp ....................   Passed    0.08 sec
        Start  12: cpu-graph-inference-int8-cpp
 12/102 Test  #12: cpu-graph-inference-int8-cpp .....................   Passed    0.11 sec
        Start  13: cpu-matmul-perf-cpp
 13/102 Test  #13: cpu-matmul-perf-cpp ..............................   Passed   15.13 sec
        Start  14: cpu-memory-format-propagation-cpp
 14/102 Test  #14: cpu-memory-format-propagation-cpp ................   Passed    0.03 sec
        Start  15: cpu-performance-profiling-cpp
 15/102 Test  #15: cpu-performance-profiling-cpp ....................   Passed    1.18 sec
        Start  16: cpu-primitives-augru-cpp
 16/102 Test  #16: cpu-primitives-augru-cpp .........................   Passed    0.03 sec
        Start  17: cpu-primitives-batch-normalization-cpp
 17/102 Test  #17: cpu-primitives-batch-normalization-cpp ...........   Passed    0.03 sec
        Start  18: cpu-primitives-binary-cpp
 18/102 Test  #18: cpu-primitives-binary-cpp ........................   Passed    0.03 sec
        Start  19: cpu-primitives-concat-cpp
 19/102 Test  #19: cpu-primitives-concat-cpp ........................   Passed    0.03 sec
        Start  20: cpu-primitives-convolution-cpp
 20/102 Test  #20: cpu-primitives-convolution-cpp ...................   Passed    0.03 sec
        Start  21: cpu-primitives-eltwise-cpp
 21/102 Test  #21: cpu-primitives-eltwise-cpp .......................   Passed    0.04 sec
        Start  22: cpu-primitives-group-normalization-cpp
 22/102 Test  #22: cpu-primitives-group-normalization-cpp ...........   Passed    0.34 sec
        Start  23: cpu-primitives-inner-product-cpp
 23/102 Test  #23: cpu-primitives-inner-product-cpp .................   Passed    0.23 sec
        Start  24: cpu-primitives-layer-normalization-cpp
 24/102 Test  #24: cpu-primitives-layer-normalization-cpp ...........   Passed    0.03 sec
        Start  25: cpu-primitives-lrn-cpp
 25/102 Test  #25: cpu-primitives-lrn-cpp ...........................   Passed    0.03 sec
        Start  26: cpu-primitives-lstm-cpp
 26/102 Test  #26: cpu-primitives-lstm-cpp ..........................   Passed    0.03 sec
        Start  27: cpu-primitives-matmul-cpp
 27/102 Test  #27: cpu-primitives-matmul-cpp ........................   Passed    0.04 sec
        Start  28: cpu-primitives-pooling-cpp
 28/102 Test  #28: cpu-primitives-pooling-cpp .......................   Passed    0.03 sec
        Start  29: cpu-primitives-prelu-cpp
 29/102 Test  #29: cpu-primitives-prelu-cpp .........................   Passed    0.04 sec
        Start  30: cpu-primitives-reduction-cpp
 30/102 Test  #30: cpu-primitives-reduction-cpp .....................   Passed    0.03 sec
        Start  31: cpu-primitives-reorder-cpp
 31/102 Test  #31: cpu-primitives-reorder-cpp .......................   Passed    0.03 sec
        Start  32: cpu-primitives-resampling-cpp
 32/102 Test  #32: cpu-primitives-resampling-cpp ....................   Passed    0.04 sec
        Start  33: cpu-primitives-shuffle-cpp
 33/102 Test  #33: cpu-primitives-shuffle-cpp .......................   Passed    0.20 sec
        Start  34: cpu-primitives-softmax-cpp
 34/102 Test  #34: cpu-primitives-softmax-cpp .......................   Passed    0.03 sec
        Start  35: cpu-primitives-sum-cpp
 35/102 Test  #35: cpu-primitives-sum-cpp ...........................   Passed    0.04 sec
        Start  36: cpu-rnn-training-f32-cpp
 36/102 Test  #36: cpu-rnn-training-f32-cpp .........................   Passed    0.88 sec
        Start  37: cpu-tutorials-matmul-matmul-quantization-cpp
 37/102 Test  #37: cpu-tutorials-matmul-matmul-quantization-cpp .....   Passed    0.03 sec
        Start  38: cpu-tutorials-matmul-sgemm-and-matmul-cpp
 38/102 Test  #38: cpu-tutorials-matmul-sgemm-and-matmul-cpp ........   Passed    0.03 sec
        Start  39: cpu-tutorials-matmul-inference-int8-matmul-cpp
 39/102 Test  #39: cpu-tutorials-matmul-inference-int8-matmul-cpp ...   Passed    0.03 sec
        Start  40: api-c
 40/102 Test  #40: api-c ............................................   Passed    0.57 sec
        Start  41: test_batch_normalization
 41/102 Test  #41: test_batch_normalization .........................   Passed    0.76 sec
        Start  42: test_binary
 42/102 Test  #42: test_binary ......................................   Passed    0.83 sec
        Start  43: test_concat
 43/102 Test  #43: test_concat ......................................   Passed    0.74 sec
        Start  44: test_concurrency
 44/102 Test  #44: test_concurrency .................................   Passed    0.82 sec
        Start  45: test_convolution_backward_data_f32
 45/102 Test  #45: test_convolution_backward_data_f32 ...............   Passed    9.24 sec
        Start  46: test_convolution_backward_weights_f32
 46/102 Test  #46: test_convolution_backward_weights_f32 ............   Passed    8.78 sec
        Start  47: test_convolution_eltwise_forward_f32
 47/102 Test  #47: test_convolution_eltwise_forward_f32 .............   Passed   11.10 sec
        Start  48: test_convolution_eltwise_forward_x8s8f32s32
 48/102 Test  #48: test_convolution_eltwise_forward_x8s8f32s32 ......   Passed    4.23 sec
        Start  49: test_convolution_forward_f32
 49/102 Test  #49: test_convolution_forward_f32 .....................   Passed    8.04 sec
        Start  50: test_convolution_forward_u8s8fp
 50/102 Test  #50: test_convolution_forward_u8s8fp ..................   Passed    0.86 sec
        Start  51: test_convolution_forward_u8s8s32
 51/102 Test  #51: test_convolution_forward_u8s8s32 .................   Passed    0.87 sec
        Start  52: test_cross_engine_reorder
 52/102 Test  #52: test_cross_engine_reorder ........................   Passed    0.69 sec
        Start  53: test_deconvolution
 53/102 Test  #53: test_deconvolution ...............................   Passed    1.02 sec
        Start  54: test_eltwise
 54/102 Test  #54: test_eltwise .....................................   Passed    0.87 sec
        Start  55: test_group_normalization
 55/102 Test  #55: test_group_normalization .........................   Passed    0.75 sec
        Start  56: test_iface_attr
 56/102 Test  #56: test_iface_attr ..................................   Passed    0.73 sec
        Start  57: test_iface_attr_quantization
 57/102 Test  #57: test_iface_attr_quantization .....................   Passed    0.69 sec
        Start  58: test_iface_binary_bcast
 58/102 Test  #58: test_iface_binary_bcast ..........................   Passed    0.69 sec
        Start  59: test_iface_handle
 59/102 Test  #59: test_iface_handle ................................   Passed    0.69 sec
        Start  60: test_iface_pd
 60/102 Test  #60: test_iface_pd ....................................   Passed    0.69 sec
        Start  61: test_iface_pd_iter
 61/102 Test  #61: test_iface_pd_iter ...............................   Passed    0.69 sec
        Start  62: test_iface_primitive_cache
 62/102 Test  #62: test_iface_primitive_cache .......................   Passed    0.69 sec
        Start  63: test_iface_runtime_dims
 63/102 Test  #63: test_iface_runtime_dims ..........................   Passed    0.69 sec
        Start  64: test_iface_weights_format
 64/102 Test  #64: test_iface_weights_format ........................   Passed    0.69 sec
        Start  65: test_iface_wino_convolution
 65/102 Test  #65: test_iface_wino_convolution ......................   Passed    0.69 sec
        Start  66: test_inner_product_backward_data
 66/102 Test  #66: test_inner_product_backward_data .................   Passed    0.79 sec
        Start  67: test_inner_product_backward_weights
 67/102 Test  #67: test_inner_product_backward_weights ..............   Passed    1.18 sec
        Start  68: test_inner_product_forward
 68/102 Test  #68: test_inner_product_forward .......................   Passed    0.81 sec
        Start  69: test_layer_normalization
 69/102 Test  #69: test_layer_normalization .........................   Passed    0.80 sec
        Start  70: test_lrn
 70/102 Test  #70: test_lrn .........................................   Passed    2.76 sec
        Start  71: test_matmul
 71/102 Test  #71: test_matmul ......................................   Passed    0.71 sec
        Start  72: test_persistent_cache_api
 72/102 Test  #72: test_persistent_cache_api ........................   Passed    0.69 sec
        Start  73: test_pooling_backward
 73/102 Test  #73: test_pooling_backward ............................   Passed    9.23 sec
        Start  74: test_pooling_forward
 74/102 Test  #74: test_pooling_forward .............................   Passed   16.49 sec
        Start  75: test_prelu
 75/102 Test  #75: test_prelu .......................................   Passed    0.75 sec
        Start  76: test_primitive_cache_mt
 76/102 Test  #76: test_primitive_cache_mt ..........................   Passed    0.72 sec
        Start  77: test_reduction
 77/102 Test  #77: test_reduction ...................................   Passed    0.73 sec
        Start  78: test_reorder
 78/102 Test  #78: test_reorder .....................................   Passed    1.40 sec
        Start  79: test_resampling
 79/102 Test  #79: test_resampling ..................................   Passed    0.76 sec
        Start  80: test_rnn_forward
 80/102 Test  #80: test_rnn_forward .................................   Passed    0.71 sec
        Start  81: test_shuffle
 81/102 Test  #81: test_shuffle .....................................   Passed    0.70 sec
        Start  82: test_softmax
 82/102 Test  #82: test_softmax .....................................   Passed    0.85 sec
        Start  83: test_sum
 83/102 Test  #83: test_sum .........................................   Passed    3.20 sec
        Start  84: test_convolution_format_any
 84/102 Test  #84: test_convolution_format_any ......................   Passed    0.69 sec
        Start  85: test_gemm_bf16bf16bf16
 85/102 Test  #85: test_gemm_bf16bf16bf16 ...........................   Passed    0.69 sec
        Start  86: test_gemm_bf16bf16f32
 86/102 Test  #86: test_gemm_bf16bf16f32 ............................   Passed    0.70 sec
        Start  87: test_gemm_f16
 87/102 Test  #87: test_gemm_f16 ....................................   Passed    0.69 sec
        Start  88: test_gemm_f16f16f32
 88/102 Test  #88: test_gemm_f16f16f32 ..............................   Passed    0.69 sec
        Start  89: test_gemm_f32
 89/102 Test  #89: test_gemm_f32 ....................................   Passed    3.63 sec
        Start  90: test_gemm_s8s8s32
 90/102 Test  #90: test_gemm_s8s8s32 ................................   Passed    7.92 sec
        Start  91: test_gemm_s8u8s32
 91/102 Test  #91: test_gemm_s8u8s32 ................................   Passed    0.71 sec
        Start  92: test_gemm_u8s8s32
 92/102 Test  #92: test_gemm_u8s8s32 ................................   Passed    7.83 sec
        Start  93: test_gemm_u8u8s32
 93/102 Test  #93: test_gemm_u8u8s32 ................................   Passed    0.71 sec
        Start  94: test_global_scratchpad
 94/102 Test  #94: test_global_scratchpad ...........................   Passed    0.69 sec
        Start  95: test_api
 95/102 Test  #95: test_api .........................................   Passed    0.71 sec
        Start  96: test_internals_env_vars_dnnl
 96/102 Test  #96: test_internals_env_vars_dnnl .....................   Passed    0.69 sec
        Start  97: test_internals_env_vars_onednn
 97/102 Test  #97: test_internals_env_vars_onednn ...................   Passed    0.69 sec
        Start  98: test_internals
 98/102 Test  #98: test_internals ...................................   Passed    0.70 sec
        Start  99: test_regression
 99/102 Test  #99: test_regression ..................................   Passed    0.72 sec
        Start 100: test_graph_api_cpu
100/102 Test #100: test_graph_api_cpu ...............................   Passed    0.50 sec
        Start 101: test_graph_unit_cpu
101/102 Test #101: test_graph_unit_cpu ..............................   Passed    6.75 sec
        Start 102: noexcept-cpp
102/102 Test #102: noexcept-cpp .....................................   Passed    0.02 sec

100% tests passed, 0 tests failed out of 102

Total Test time (real) = 162.52 sec
```

### CODE REVIEW

File: mingw-w64-onednn/001-aarch64-mingw.patch [hunk 1]
```diff
@@ -0,0 +1,75 @@
+diff -bur oneDNN-3.3.2-orig/src/cpu/aarch64/cpu_isa_traits.hpp oneDNN-3.3.2/src/cpu/aarch64/cpu_isa_traits.hpp
+--- oneDNN-3.3.2-orig/src/cpu/aarch64/cpu_isa_traits.hpp	2023-12-03 02:20:17.434762400 -0700
++++ oneDNN-3.3.2/src/cpu/aarch64/cpu_isa_traits.hpp	2023-12-03 02:24:46.795841200 -0700
+@@ -28,7 +28,7 @@
+ 
+ /* in order to make selinux happy memory that would be marked with X-bit should
+  * be obtained with mmap */
+-#if !defined(_WIN32)
++#if !defined(_WIN32) && !defined(__MINGW64__)
```
Comment (Biswa96) [inline mingw-w64-onednn/001-aarch64-mingw.patch:9]: Doesn't `!_WIN32` disable mingw also?

Comment (hmartinez82) [reply mingw-w64-onednn/001-aarch64-mingw.patch:9]: Yes, but that's the intent :)

File: mingw-w64-onednn/001-aarch64-mingw.patch [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ #include <sys/mman.h>
+diff -bur oneDNN-3.3.2-orig/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_code_array.h oneDNN-3.3.2/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_code_array.h
+--- oneDNN-3.3.2-orig/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_code_array.h	2023-12-03 02:20:17.517356400 -0700
++++ oneDNN-3.3.2/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_code_array.h	2023-12-03 02:29:02.393195200 -0700
+@@ -20,7 +20,7 @@
+ static const size_t CSIZE = sizeof(uint32_t);
+ 
+ inline void *AlignedMalloc(size_t size, size_t alignment) {
+-#ifdef _MSC_VER
++#if defined(_MSC_VER) || defined(__MINGW64__)
```
Comment (Biswa96) [inline mingw-w64-onednn/001-aarch64-mingw.patch:42]: This simplifies to `#ifdef _WIN32`.

File: mingw-w64-onednn/001-aarch64-mingw.patch (hunk 1 above)
Comment (hmartinez82) [reply mingw-w64-onednn/001-aarch64-mingw.patch:9]: And this patch is only for aarch. I'm thinking about upstreaming this patch. Their code seems to be handling MSVC for ARM64 already. So it just needs to stop trying to use nmap (like it does when building with MSVC) in MINGW.

File: mingw-w64-onednn/001-aarch64-mingw.patch (hunk 2 above)
Comment (hmartinez82) [reply mingw-w64-onednn/001-aarch64-mingw.patch]: Yes. In many other places actually. That's what I did with the upstream patch. I'm going to use that instead :)

**Comment by hmartinez82** [comment]
@Biswa96 sorry for the late reply to came back to this. It should be good to go.

**Comment by Biswa96** [comment]
Will the patch file change in future? You can copy the patch file here also.

**Comment by hmartinez82** [comment]
Well, something happened. the .patch file is being downloaded with an old content of the upstream PR. I don't know why. Investigating now 😑

**Comment by hmartinez82** [comment]
@Biswa96 Fixed, it's now using _WIN32 as you suggested

**Review by Biswa96 — APPROVED** [review]

**Pull request merged**